File: BOJ/♠ 2000~2999/2294_동전_2.py
coin, goal = map(int, input().split())
inf = 1000000
dp = [inf] * (goal + 1)
dp[0] = 0

for i in range(coin):
    current = int(input())
    for j in range(current, goal + 1):
        dp[j] = min(dp[j], dp[j - current] + 1)

if dp[-1] == inf:
    print(-1)
else:
    print(dp[-1])

File: BOJ/♠ 2000~2999/2294_동전_2.cpp
#include <iostream>
#include <algorithm>
#define INF 10001
using namespace std;

int dp[10001];

int main()
{
    cin.tie(0)->sync_with_stdio(0);

    int coin, goal, current;
    cin >> coin >> goal;

    fill(dp, dp + 10001, INF);
    dp[0] = 0;

    for (int i = 0; i < coin; i++)
    {
        cin >> current;

        for (int j = current; j <= goal; j++)
            dp[j] = min(dp[j], dp[j - current] + 1);
    }

    if (dp[goal] == INF)
        cout << -1;
    else
        cout << dp[goal];
}